haddock-api/src/Haddock/Backends/Annot.hs

```diff
@@ -0,0 +1,185 @@
+        ns     = getNames src tcm
+        tm     = M.fromList ns
+        typs s = case s `M.lookup` tm of
+                   Nothing    -> []
+                   Just (x,_) -> [x]
+
+canonize :: (Ord b, Eq a) => [(b, (t, [a]))] -> [(b, (t, [a]))]
+canonize anns = maximumBy cmp $ groupWith fst anns
+  where maximumBy o = catMaybes . fmap (safeHead . L.sortBy o)
+        safeHead [] = Nothing
+        safeHead (x:_) = Just x
+        cmp (_,(_,x1)) (_,(_,x2))
+          | x1 == x2              = EQ
+          | length x1 < length x2 = GT
+          | otherwise             = LT
+
+getLocEs ::  (Data a) => a -> [(HsExpr Id, Loc, String)]
+getLocEs z = [(e, l, stripParens $ unsafePpr e) | L l e <- findLEs z]
+  where stripParens ('(':s)  = stripParens s
+        stripParens s        = stripRParens (reverse s)
+        stripRParens (')':s) = stripRParens s
+        stripRParens s       = reverse s
+
+getNames ::  (Data a) => FastString -> a -> [(String, (Id, Maybe Loc))]
+getNames src z = [(unsafePpr x, (x, idLoc src x)) | x <- findIds z]
+
+renderId :: Id -> String
+renderId = showSDocForUser unsafeGlobalDynFlags neverQualify . pprTyThing showSub . AnId
+
+showSub :: ShowSub
+showSub = ShowSub {
+    ss_how_much = ShowHeader (AltPpr Nothing)
+  , ss_forall = ShowForAllWhen
+  }
+
+idLoc :: FastString -> Id -> Maybe Loc
+idLoc src x
+  | not (isGoodSrcSpan sp)
+  = Nothing
+  | Just src /= fmap srcSpanFile (realSrcSpan sp)
+  = Nothing
+  | otherwise              = Just sp
+  where sp  = nameSrcSpan $ idName x
+
+realSrcSpan :: SrcSpan -> Maybe RealSrcSpan
+realSrcSpan ss =
+  case ss of
+    RealSrcSpan rs -> Just rs
+    UnhelpfulSpan _msg -> Nothing
+
+unsafePpr :: Outputable a => a -> String
+unsafePpr =
+  showSDocUnsafe . ppr
+
+---------------------------------------------------------------------------
+-- Visiting and Extracting Identifiers ------------------------------------
+-- From Tamar Christina: http://mistuke.wordpress.com/category/vsx --------
+---------------------------------------------------------------------------
+
+type GenericQ r = forall a. Data a => a -> r
+
+-- | Summarise all nodes in top-down, left-to-right order
+everythingButQ :: (r -> r -> r) -> [TypeRep] -> GenericQ r -> GenericQ r
+everythingButQ k q f x
+  = foldl k (f x) fsp
+    where fsp = case isPost x q of
+                  True  -> []
+                  False -> gmapQ (everythingButQ k q f) x
+
+isPost :: Typeable a => a -> [TypeRep] -> Bool
+isPost a = or . map (== typeOf a)
+
+-- | Get a list of all entities that meet a predicate
+listifyBut :: Typeable r => (r -> Bool) -> [TypeRep] -> GenericQ [r]
+listifyBut p q
+  = everythingButQ (++) q ([] `mkQ` (\x -> if p x then [x] else []))
+
+-- | Types to avoid inspecting. We should not enter any PostTcKind
+-- nor NameSet because these are blank after type checking.
+skipGuards :: [TypeRep]
+skipGuards = [ typeRep (Proxy :: Proxy NameSet)
+             , typeRep (Proxy :: Proxy (PostTc Id Kind))]
```

Comment:
I'm not totally sure if this is still necessary - I saw no ill effect from turning these off. The rationale behind them is in Tamar's 2011 blog post linked above.

Reply:
This is ok I guess.